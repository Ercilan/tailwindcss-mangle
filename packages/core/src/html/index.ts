import { parse, serialize } from 'parse5'
import { traverse } from '@parse5/tools'
import { IHtmlHandlerOptions } from '../types'

export function htmlHandler(rawSource: string, options: IHtmlHandlerOptions) {
  const { runtimeSet, classGenerator } = options
  const fragment = parse(rawSource)
  traverse(fragment, {
    element(node, parent) {
      const attr = node.attrs.find((x) => x.name === 'class')
      if (attr) {
        const arr = attr.value.split(/\s/).filter((x) => x)
        attr.value = arr
          .map((x) => {
            if (runtimeSet.has(x)) {
              return classGenerator.generateClassName(x).name
            }
            return x
          })
          .join(' ')
      }
    }
  })
  return serialize(fragment)
}
