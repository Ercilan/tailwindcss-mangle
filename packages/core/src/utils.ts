import micromatch from 'micromatch'

const { isMatch } = micromatch

export function createGlobMatcher(pattern: string | string[] | undefined, fallbackValue: boolean = false) {
  if (pattern === undefined) {
    return function () {
      return fallbackValue
    }
  }
  return function (file: string) {
    return isMatch(file, pattern)
  }
}

export { defaultMangleClassFilter, isMap, isRegexp } from '@tailwindcss-mangle/shared'
