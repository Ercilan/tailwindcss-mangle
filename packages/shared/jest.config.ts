import type { Config } from 'jest'
import baseConfig from '../../jest.config'
const config: Config = {
  projects: [
    {
      ...baseConfig
      // transformIgnorePatterns: ['/node_modules/(?!(@parse5/)/tools)']
    }
  ],
  collectCoverage: true,
  coverageDirectory: '../../coverage/tailwindcss-mangle-shared',
  coverageProvider: 'v8',
  coveragePathIgnorePatterns: ['/node_modules/', '/test/']
}

export default config
