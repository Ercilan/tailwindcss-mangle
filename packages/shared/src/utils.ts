import type { IClassGeneratorOptions, IClassGenerator } from './types'

export const defaultMangleClassFilter = (className: string) => {
  // ignore className like 'filter','container'
  // it may be dangerous to mangle/rename all StringLiteral , so use /-/ test for only those with /-/ like:
  // bg-[#123456] w-1 etc...
  return /[:-]/.test(className)
}

export function groupBy<T>(arr: T[], cb: (arg: T) => string): Record<string, T[]> {
  if (!Array.isArray(arr)) {
    throw new TypeError('expected an array for first argument')
  }

  if (typeof cb !== 'function') {
    throw new TypeError('expected a function for second argument')
  }

  const result: Record<string, T[]> = {}
  for (const item of arr) {
    const bucketCategory = cb(item)
    const bucket = result[bucketCategory]

    if (Array.isArray(bucket)) {
      result[bucketCategory].push(item)
    } else {
      result[bucketCategory] = [item]
    }
  }

  return result
}

export const acceptChars = [...'abcdefghijklmnopqrstuvwxyz']

export function stripEscapeSequence(words: string) {
  return words.replaceAll('\\', '')
}

export const validate = (opts: IClassGeneratorOptions, classGenerator: IClassGenerator) => {
  if (!opts.log) return
  for (const className in classGenerator.newClassMap) {
    const c = classGenerator.newClassMap[className]
    if (c.usedBy.length > 0) {
      continue
    }
    if (/.+\.css:*$/.test(c.usedBy[0])) {
      console.log(`The class name '${className}' is not used: defined at ${c.usedBy[0]}.`)
    } else {
      console.log(`The class name '${className}' is not defined: used at ${c.usedBy[0]}.`)
    }
  }
}

export function isRegexp(value: unknown) {
  return Object.prototype.toString.call(value) === '[object RegExp]'
}

export function isMap(value: unknown) {
  return Object.prototype.toString.call(value) === '[object Map]'
}

export function regExpTest(arr: (string | RegExp)[] = [], str: string) {
  if (Array.isArray(arr)) {
    for (const item of arr) {
      if (typeof item === 'string') {
        if (item === str) {
          return true
        }
      } else if (isRegexp(item)) {
        item.lastIndex = 0
        if (item.test(str)) {
          return true
        }
      }
    }
    return false
  }
  throw new TypeError("paramater 'arr' should be a Array of Regexp | String !")
}
