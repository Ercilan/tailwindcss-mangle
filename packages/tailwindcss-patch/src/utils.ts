import fss from 'node:fs'
import fs from 'node:fs/promises'
import path from 'node:path'
import type { SyncOpts } from 'resolve'
import pkg from 'resolve'
import type { PackageJson } from 'pkg-types'

const { sync } = pkg

export function ensureFileContent(filepaths: string | string[]) {
  if (typeof filepaths === 'string') {
    filepaths = [filepaths]
  }
  let content
  for (const filepath of filepaths) {
    if (fss.existsSync(filepath)) {
      content = fss.readFileSync(filepath, {
        encoding: 'utf8',
      })
      break
    }
  }
  return content
}

export function requireResolve(id: string, opts?: SyncOpts) {
  return sync(id, opts)
}

export async function ensureDir(p: string) {
  try {
    await fs.access(p)
  }
  catch {
    await fs.mkdir(p, {
      recursive: true,
    })
  }
}

function searchPackageJSON(dir: string) {
  let packageJsonPath
  while (true) {
    if (!dir) {
      return
    }
    const newDir = path.dirname(dir)
    if (newDir === dir) {
      return
    }
    dir = newDir
    packageJsonPath = path.join(dir, 'package.json')
    if (fss.existsSync(packageJsonPath)) {
      break
    }
  }

  return packageJsonPath
}

function getTailwindcssEntry(name: string = 'tailwindcss', opts?: SyncOpts) {
  return requireResolve(name, opts)
}

function getPackageJsonPath(name: string, options: SyncOpts = {}) {
  const entry = getTailwindcssEntry(name, options)
  if (!entry) {
    return
  }

  return searchPackageJSON(entry)
}

export function getPackageInfoSync(name: string, options: SyncOpts = {}) {
  const packageJsonPath = getPackageJsonPath(name, options)
  if (!packageJsonPath) {
    return
  }

  const packageJson: PackageJson = JSON.parse(fss.readFileSync(packageJsonPath, 'utf8'))

  return {
    name,
    version: packageJson.version,
    rootPath: path.dirname(packageJsonPath),
    packageJsonPath,
    packageJson,
  }
}

export function isObject(val: any) {
  return val != null && typeof val === 'object' && Array.isArray(val) === false
};
