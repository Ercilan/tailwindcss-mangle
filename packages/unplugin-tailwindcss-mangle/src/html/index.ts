import { parse, serialize } from 'parse5'
import { traverse } from '@parse5/tools'
import { IHandlerOptions } from '@/types'

export function htmlHandler(rawSource: string, options: IHandlerOptions) {
  const { runtimeSet, classGenerator } = options
  const fragment = parse(rawSource)
  traverse(fragment, {
    element(node, parent) {
      const attr = node.attrs.find((x) => x.name === 'class')
      if (attr) {
        const arr = attr.value.split(/\s/).filter((x) => x)
        attr.value = arr
          .map((x) => {
            if (runtimeSet.has(x)) {
              return classGenerator.generateClassName(x).name
            }
            return x
          })
          .join(' ')
      }
    }
  })
  return serialize(fragment)
}
