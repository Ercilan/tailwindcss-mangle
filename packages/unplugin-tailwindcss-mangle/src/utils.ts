import type { IClassGeneratorOptions, IClassGenerator } from './types'
import micromatch from 'micromatch'
import fs from 'fs'
import path from 'path'

import { pluginName } from './constants'
const { isMatch } = micromatch

export const isMangleClass = (className: string) => {
  // ignore className like 'filter','container'
  // it may be dangerous to mangle/rename all StringLiteral , so use /-/ test for only those with /-/ like:
  // bg-[#123456] w-1 etc...
  return /[-:]/.test(className)
}

export function groupBy<T>(arr: T[], cb: (arg: T) => string): Record<string, T[]> {
  if (!Array.isArray(arr)) {
    throw new Error('expected an array for first argument')
  }

  if (typeof cb !== 'function') {
    throw new Error('expected a function for second argument')
  }

  const result: Record<string, T[]> = {}
  for (let i = 0; i < arr.length; i++) {
    const item = arr[i]
    const bucketCategory = cb(item)
    const bucket = result[bucketCategory]

    if (!Array.isArray(bucket)) {
      result[bucketCategory] = [item]
    } else {
      result[bucketCategory].push(item)
    }
  }

  return result
}

export function getGroupedEntries<T>(
  entries: [string, T][],
  options = {
    cssMatcher(file: string) {
      return /\.css$/.test(file)
    },
    htmlMatcher(file: string) {
      return /\.html?$/.test(file)
    },
    jsMatcher(file: string) {
      return /\.[cm]?js$/.test(file)
    }
  }
) {
  const { cssMatcher, htmlMatcher, jsMatcher } = options
  const groupedEntries = groupBy(entries, ([file]) => {
    if (cssMatcher(file)) {
      return 'css'
    } else if (htmlMatcher(file)) {
      return 'html'
    } else if (jsMatcher(file)) {
      return 'js'
    } else {
      return 'other'
    }
  })
  if (!groupedEntries.css) {
    groupedEntries.css = []
  }
  if (!groupedEntries.html) {
    groupedEntries.html = []
  }
  if (!groupedEntries.js) {
    groupedEntries.js = []
  }
  if (!groupedEntries.other) {
    groupedEntries.other = []
  }
  return groupedEntries as Record<'css' | 'html' | 'js' | 'other', [string, T][]>
}

export const acceptChars = 'abcdefghijklmnopqrstuvwxyz'.split('')

export function stripEscapeSequence(words: string) {
  return words.replace(/\\/g, '')
}

export const validate = (opts: IClassGeneratorOptions, classGenerator: IClassGenerator) => {
  if (!opts.log) return
  for (const className in classGenerator.newClassMap) {
    const c = classGenerator.newClassMap[className]
    if (c.usedBy.length >= 1) {
      continue
    }
    if (c.usedBy[0].match(/.+\.css:*$/)) {
      console.log(`The class name '${className}' is not used: defined at ${c.usedBy[0]}.`)
    } else {
      console.log(`The class name '${className}' is not defined: used at ${c.usedBy[0]}.`)
    }
  }
}

export function isRegexp(value: unknown) {
  return Object.prototype.toString.call(value) === '[object RegExp]'
}

export function isMap(value: unknown) {
  return Object.prototype.toString.call(value) === '[object Map]'
}

export function regExpTest(arr: (string | RegExp)[] = [], str: string) {
  if (Array.isArray(arr)) {
    for (let i = 0; i < arr.length; i++) {
      const item = arr[i]
      if (typeof item === 'string') {
        if (item === str) {
          return true
        }
      } else if (isRegexp(item)) {
        item.lastIndex = 0
        if (item.test(str)) {
          return true
        }
      }
    }
    return false
  }
  throw new TypeError("paramater 'arr' should be a Array of Regexp | String !")
}

export function escapeStringRegexp(str: string) {
  if (typeof str !== 'string') {
    throw new TypeError('Expected a string')
  }
  return str.replace(/[|\\{}()[\]^$+*?.]/g, '\\$&').replace(/-/g, '\\x2d')
}

export function createGlobMatcher(pattern: string | string[] | undefined, fallbackValue: boolean = false) {
  if (typeof pattern === 'undefined') {
    return function (file: string) {
      return fallbackValue
    }
  }
  return function (file: string) {
    return isMatch(file, pattern)
  }
}

export function getCacheDir(basedir = process.cwd()) {
  return path.resolve(basedir, 'node_modules/.cache', pluginName)
}

export function mkCacheDirectory(cwd = process.cwd()) {
  const cacheDirectory = getCacheDir(cwd)

  const exists = fs.existsSync(cacheDirectory)
  if (!exists) {
    fs.mkdirSync(cacheDirectory, {
      recursive: true
    })
  }
  return cacheDirectory
}

export function cacheDump(filename: string, data: any[] | Set<any>, basedir?: string) {
  try {
    const dir = mkCacheDirectory(basedir)
    fs.writeFileSync(path.resolve(dir, filename), JSON.stringify(Array.from(data), null, 2), 'utf-8')
  } catch (error) {
    console.log(error)
  }
}
