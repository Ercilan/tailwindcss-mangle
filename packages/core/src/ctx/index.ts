import type { ClassMapOutputItem, MangleUserConfig } from '@tailwindcss-mangle/config'
import process from 'node:process'
import { defaultMangleClassFilter, escapeStringRegexp } from '@/utils'
import { getConfig } from '@tailwindcss-mangle/config'
import { ClassGenerator } from '@tailwindcss-mangle/shared'
import { defu } from 'defu'
import { sort } from 'fast-sort'
import fs from 'fs-extra'
import { dirname, isAbsolute, resolve } from 'pathe'

interface InitConfigOptions {
  cwd?: string
  classList?: string[]
  mangleOptions?: MangleUserConfig
}

export class Context {
  options: MangleUserConfig
  public replaceMap: Map<string, string>
  classSet: Set<string>

  classGenerator: ClassGenerator

  preserveFunctionSet: Set<string>
  preserveClassNamesSet: Set<string>
  preserveFunctionRegexs: RegExp[]
  constructor() {
    this.options = {}
    this.classSet = new Set()
    this.replaceMap = new Map()
    this.classGenerator = new ClassGenerator()
    this.preserveFunctionSet = new Set()
    this.preserveClassNamesSet = new Set()
    this.preserveFunctionRegexs = []
  }

  isPreserveClass(className: string) {
    return this.preserveClassNamesSet.has(className)
  }

  addPreserveClass(className: string) {
    return this.preserveClassNamesSet.add(className)
  }

  isPreserveFunction(calleeName: string) {
    return this.preserveFunctionSet.has(calleeName)
  }

  private mergeOptions(...opts: (MangleUserConfig | undefined)[]) {
    // 配置选项优先
    this.options = defu(this.options, ...opts)
    this.classGenerator = new ClassGenerator(this.options.classGenerator)
    this.preserveFunctionSet = new Set(this.options?.preserveFunction ?? [])
    this.preserveFunctionRegexs = [...this.preserveFunctionSet.values()].map((x) => {
      return new RegExp(`${escapeStringRegexp(x)}\\(([^)]*)\\)`, 'g')
    })
  }

  currentMangleClassFilter(className: string) {
    return (this.options.mangleClassFilter ?? defaultMangleClassFilter)(className)
  }

  getClassSet() {
    return this.classSet
  }

  getReplaceMap() {
    return this.replaceMap // map
  }

  addToUsedBy(key: string, file?: string) {
    if (!file) {
      return
    }
    const hit = this.classGenerator.newClassMap[key]
    if (hit) {
      hit.usedBy.add(file)
    }
  }

  loadClassSet(classList: string[]) {
    const list = sort(classList).desc(c => c.length)
    for (const className of list) {
      if (this.currentMangleClassFilter(className)) {
        this.classSet.add(className)
      }
    }
  }

  async initConfig(opts: InitConfigOptions = {}) {
    const { cwd, classList: _classList, mangleOptions } = opts
    const { config, cwd: configCwd } = await getConfig(cwd)
    if (mangleOptions?.classMapOutput === true) {
      mangleOptions.classMapOutput = config.mangle?.classMapOutput
      if (typeof mangleOptions.classMapOutput === 'object') {
        mangleOptions.classMapOutput.enable = true
      }
    }
    this.mergeOptions(mangleOptions, config?.mangle)
    if (_classList) {
      this.loadClassSet(_classList)
    }
    else {
      let jsonPath = this.options.classListPath ?? resolve(process.cwd(), config?.patch?.output?.filename as string)
      if (!isAbsolute(jsonPath)) {
        jsonPath = resolve(configCwd ?? process.cwd(), jsonPath)
      }

      if (jsonPath && fs.existsSync(jsonPath)) {
        const rawClassList = fs.readFileSync(jsonPath, 'utf8')
        const list = JSON.parse(rawClassList) as string[]
        // why?
        // cause bg-red-500 and bg-red-500/50 same time
        // transform bg-red-500/50 first
        this.loadClassSet(list)
      }
    }

    for (const cls of this.classSet) {
      this.classGenerator.generateClassName(cls)
    }

    for (const x of Object.entries(this.classGenerator.newClassMap)) {
      this.replaceMap.set(x[0], x[1].name)
    }
    return config
  }

  async dump() {
    try {
      const arr = Object.entries(this.classGenerator.newClassMap).map<ClassMapOutputItem>((x) => {
        return {
          before: x[0],
          after: x[1].name,
          usedBy: Array.from(x[1].usedBy),
        }
      })
      if (typeof this.options.classMapOutput === 'function') {
        await this.options.classMapOutput(arr)
      }
      else if (typeof this.options.classMapOutput === 'object' && this.options.classMapOutput.enable && this.options.classMapOutput.filename) {
        fs.mkdirSync(dirname(this.options.classMapOutput.filename), { recursive: true })
        fs.writeFileSync(this.options.classMapOutput.filename, JSON.stringify(arr, null, 2))
      }
    }
    catch (error) {
      console.error(`[tailwindcss-mangle]: ${error}`)
    }
  }
}
