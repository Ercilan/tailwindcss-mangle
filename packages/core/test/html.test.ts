import { splitCode } from '@tailwindcss-mangle/shared'
import { getTestCase } from './utils'
import { htmlHandler } from '@/html'
import { Context } from '@/ctx'

describe('html handler', () => {
  // let classGenerator: ClassGenerator
  let ctx: Context
  beforeEach(() => {
    // classGenerator = new ClassGenerator()
    ctx = new Context()
  })
  it('common usage', () => {
    const replaceMap = ctx.replaceMap

    for (const x of splitCode('text-3xl font-bold underline')) {
      replaceMap.set(x, true)
    }
    const res = htmlHandler(getTestCase('hello-world.html'), {
      ctx,
    })
    expect(res).toMatchSnapshot()
  })

  it('trailing slash case', () => {
    const replaceMap = ctx.replaceMap

    for (const x of splitCode('bg-red-500 bg-red-500/50')) {
      replaceMap.set(x, true)
    }
    const res = htmlHandler(getTestCase('trailing-slash.html'), {
      ctx,
    })
    expect(res).toMatchSnapshot()
  })

  it('trailing slash case 0', () => {
    const replaceMap = ctx.replaceMap

    for (const x of splitCode('bg-red-500 bg-red-500/50')) {
      replaceMap.set(x, true)
    }
    const res = htmlHandler(getTestCase('trailing-slash-0.html'), {
      ctx,
    })
    expect(res).toMatchSnapshot()
  })
})
