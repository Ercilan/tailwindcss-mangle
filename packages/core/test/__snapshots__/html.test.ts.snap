// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`html handler > common usage 1`] = `
"<!DOCTYPE html><html><head>
  <meta charset=\\"UTF-8\\">
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
  <link href=\\"/main.css\\" rel=\\"stylesheet\\">
</head>

<body>
  <h1 class=\\"tw-a tw-b tw-c\\">
    Hello world!
  </h1>


</body></html>"
`;

exports[`html handler > trailing slash case 0 1`] = `
"<!DOCTYPE html><html lang=\\"en\\"><head>
  <meta charset=\\"UTF-8\\">
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
  <title>Document</title>
</head>

<body>
  <div>
    <div class=\\"tw-a\\">
      Block with bg-red-500/50 class
    </div>
    <div class=\\"tw-b\\">
      Block with bg-red-500 class
    </div>
  </div>


</body></html>"
`;

exports[`html handler > trailing slash case 1`] = `
"<!DOCTYPE html><html lang=\\"en\\"><head>
  <meta charset=\\"UTF-8\\">
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
  <title>Document</title>
</head>

<body>
  <div>
    <div class=\\"tw-a\\">
      Block with bg-red-500 class
    </div>
    <div class=\\"tw-b\\">
      Block with bg-red-500/50 class
    </div>
  </div>


</body></html>"
`;
